### ISSUE TITLE
Optimizer within Estimator computes incorrectly small gradient when used with MirroredStrategy

### ISSUE THREAD

**Issue by lightvector** [tensorflow/tensorflow#35333]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes.
- OS Platform and Distribution: Linux Ubuntu 18.04
- TensorFlow installed from: Anaconda, clean install: `conda create -n tf2 tensorflow-gpu=2.0.0`
- TensorFlow version: 2.0.0
- Python version: 3.7.5
- CUDA/cuDNN version: 10.0.130, 7.6.4
- GPU model and memory: Tesla V100, 16GB

**Describe the current behavior**
A gradient is reported that is only half of the expected gradient when using distributed training with Estimator. See example code below.
**Describe the expected behavior**
The gradient should be twice as large. If somehow this behavior is actually intended, then this needs to be much more loudly documented since it is quite unexpected - right now I cannot find any place where it is documented at all. For example, https://www.tensorflow.org/guide/distributed_training#using_tfdistributestrategy_with_estimator_limited_support and https://www.tensorflow.org/api_docs/python/tf/compat/v1/train/Optimizer#compute_gradients both have no mention of it.

**Code to reproduce the issue**
```
import sys
import os
import tensorflow as tf
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '1'

def model_fn(features,labels,mode,params):
  if mode == tf.estimator.ModeKeys.PREDICT:
    raise NotImplementedError()
  if mode == tf.estimator.ModeKeys.EVAL:
    raise NotImplementedError()
  if mode == tf.estimator.ModeKeys.TRAIN:
    w = tf.compat.v1.get_variable(initializer=tf.zeros([2]),name="w")
    losses = tf.square(w - labels)
    loss = tf.reduce_sum(losses)
    optimizer = tf.compat.v1.train.GradientDescentOptimizer(learning_rate=0.01)
    gradients = optimizer.compute_gradients(tf.reduce_sum(loss))
    global_step=tf.compat.v1.train.get_global_step()
    train_op = optimizer.apply_gradients(gradients,global_step=global_step)
    print_op = tf.print(
      w,
      losses,
      [(grad,var.name) for (grad,var) in gradients],
    )
    return tf.estimator.EstimatorSpec(
      mode,
      loss=loss,
      train_op=tf.group(train_op,print_op)
    )

def input_fn():
  dataset = tf.data.Dataset.from_tensors(({}, [1.,2.]))
  return dataset.repeat(1)

print("SINGLE GPU")
estimator = tf.estimator.Estimator(
  model_fn=model_fn,
  model_dir="test_single_gpu/",
  params={}
)
estimator.train(input_fn)

def input_fn():
  dataset = tf.data.Dataset.from_tensors(({}, [1.,2.]))
  return dataset.repeat(2)

print("MULTI GPU")
strategy = tf.distribute.MirroredStrategy(devices=["/GPU:0","/GPU:1"])
estimator = tf.estimator.Estimator(
  model_fn=model_fn,
  model_dir="test_multi_gpu/",
  params={},
  config=tf.estimator.RunConfig(train_distribute = strategy)
)
estimator.train(input_fn)
```

**Other info / logs**
The relevant portion of the output of the above test case is:
```
SINGLE GPU
[0 0] [1 4] [([-2 -4], 'w:0')]
MULTI GPU
[0 0] [1 4] [([-1 -2], 'w:0')]
[0 0] [1 4] [([-1 -2], 'w:0')]
```
The model is attempting to fit two weights to be equal to 1 and 2, respectively, penalizing them by the squared error. The weights are initialized to 0.
* Single GPU case: We can see it reads one data element from the dataset, and correctly computes the loss as as "1" and "4", which are the squared differences. The gradients are correctly computed as "-2" and "-4", which indeed the mathematical derivatives of (x-1)^2 and (x-2)^2 at x = 0, respectively. So this is all correct.
* Multi GPU case: We can see each GPU independently and in parallel reads an element from the dataset, reading two elements in total. Each one correctly computes the loss as "1" and "4" as before on its own element. However, each one separately only computes "-1" and "-2" as the gradient. This is wrong, each one is a factor of 2 too small.

(edit: corrected linux version, some grammar edits)

**Comment by guptapriya** [comment]
@lightvector - yes this is expected - for multi GPU, the loss is scaled by number of replicas[1], and later on in the optimizer, the gradients are summed over the replicas[2]. Overall this gives the correct averaged gradients. Another alternative would be to not scale the loss, but average the gradients. This would be equivalent but we found this to be slower in performance. 

You're right that we don't seem to document the behavior with estimator & old optimizers anywhere. We do document the behavior in new optimizers[3]. Happy to accept a PR to add this documentation in optimizer v1 if you want to send one.

[1] https://github.com/tensorflow/tensorflow/blob/aa8a34071d1170b6361d11676f3256c3311af2d1/tensorflow/python/training/optimizer.py#L468
[2] https://github.com/tensorflow/tensorflow/blob/aa8a34071d1170b6361d11676f3256c3311af2d1/tensorflow/python/training/optimizer.py#L668
[3] https://github.com/tensorflow/tensorflow/blob/aa8a34071d1170b6361d11676f3256c3311af2d1/tensorflow/python/keras/optimizer_v2/optimizer_v2.py#L151

**Comment by lightvector** [comment]
This seems still incorrect though - how can this actually be intended?

It is natural to expect that the numeric output of "compute the gradient" is the gradient. At the point of computing the gradient, there is no inter-GPU communication that should be happening, right? So there is no question of taking the mean versus the sum. Each GPU independently computes the gradient on its own batch, and if the user then asks what the gradient on that batch, they should get the gradient on that batch. 

This matters in some cases. For example, if using tf.clip_by_global_norm, one might expect to specify a bound on the global norm of the per-replica batch - expecting that each GPU independently will perform the clipping on its batch, then jointly apply the results as usual. If the per-replica batch was obtained by splitting a global batch size (e.g. using dataset.unbatch()) into two, then one might naturally lower the norm clipping threshold by a factor of sqrt(2), since a batch of uncorrelated data samples split into two replica-batches will result in gradients sqrt(2) smaller on each replica-batch on average.

Instead, one has to specify a bound that is sqrt(8) times smaller, because not only are the gradients on each per-replica batch naturally smaller, but Tensorflow additionally reports them as a further factor of two smaller than they truly are on that per-replica batch.

If scaling should happen, it should be entirely contained within the apply_gradients function, right? I.e. the scaling should happen at the time of applying the gradient, rather than contaminating the value that compute_gradients returns and making it no longer the gradient? Because only at the point of actually _applying_ the gradient do you now actually get cross-replica communication and only then does the question of averaging versus summing come into play. Along a note in the documentation, this would make everything far more intuitive. Otherwise you end up with the bizarre situation where "optimizer.compute_gradients" actually does not return the gradient, it returns the gradient scaled by something depending on the existence of other GPUs, even though the gradient computation should be entirely GPU-local?

**Comment by sushreebarsa** [comment]
Was able to replicate the issue in TF v2.5 ,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/3f41744c174a9bed0d9b15ea08ac7cb4/untitled120.ipynb)..Thanks !

**Comment by SuryanarayanaY** [comment]
I was able to replicate the issue in TF2.10.Please find the attached [gist](https://colab.research.google.com/gist/SuryanarayanaY/9c9885fbe5713c20da653eb8282e425a/35333.ipynb#scrollTo=6BAOjh2nlU2Q).
Thankyou!
